### PR TITLE
Make it possible to codesign, create DMG and PKG installers when cross-compiling Brave for macOS in Linux

### PR DESCRIPTION
**Pull request by mherrmann**
For https://github.com/brave/devops/issues/7918. Before this PR, it was already possible to compile Brave. Now, it is also possible to codesign, produce DMG and PKG installers.

This is especially useful for developers like me who like to work in Linux on their Macs. More information in https://github.com/brave/devops/wiki/Cross-compiling.

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

N/A. This is purely a dev concern.

### CODE REVIEW

**Review by fmarier — APPROVED** [review]
`script/brave_license_helper.py` changes look good

File: build/mac/BUILD.gn [hunk 1]
```diff
@@ -392,7 +392,12 @@ action("sign_pkg") {
 
 action("create_dmg") {
   script = "//build/gn_run_binary.py"
-  shell_script = "//chrome/installer/mac/pkg-dmg"
+
+  if (host_os == "mac") {
+    shell_script = "//chrome/installer/mac/pkg-dmg"
+  } else {
+    shell_script = "cross-compile/pkg-dmg-linux.py"
```
Comment (bridiver) [inline build/mac/BUILD.gn:399]: this should just be `script`

File: DEPS [hunk 2]
```diff
@@ -21,6 +21,10 @@ deps = {
   "third_party/ethash/src": "https://github.com/chfast/ethash.git@e4a15c3d76dc09392c7efd3e30d84ee3b871e9ce",
   "third_party/bitcoin-core/src": "https://github.com/bitcoin/bitcoin.git@8105bce5b384c72cf08b25b7c5343622754e7337", # v25.0
   "third_party/argon2/src": "https://github.com/P-H-C/phc-winner-argon2.git@62358ba2123abd17fccf2a108a301d4b52c01a7c",
+  "third_party/libdmg-hfsplus": {
```
Comment (bridiver) [inline DEPS:24]: is this the only way to get this python module? I think there's a better method than adding them to deps cc @goodov

File: build/commands/lib/config.js [hunk 3]
```diff
@@ -1221,6 +1221,12 @@ Object.defineProperty(Config.prototype, 'defaultOptions', {
       ['build'],
       ['third_party', 'depot_tools'],
     ]
+    if (this.getTargetOS() === 'mac' && process.platform !== 'darwin') {
```
[the remark this answers is not in the document]
Comment (mherrmann) [reply build/commands/lib/config.js:1224]: The directories that get added to the PYTHONPATH and normal PATH here overwrite and provide built-ins that are normally only available on macOS. If we were to add the directories unconditionally, then the overwritten implementations would take effect on macOS, and because it is not written for that OS, not work.

So in short, if we were to add the directories unconditionally, then the build on macOS would break. The `if` guard prevents this.

File: DEPS (hunk 2 above)
Comment (mherrmann) [reply DEPS:24]: This is not a Python module. It's a tool that's written in C.

File: DEPS [hunk 4]
```diff
@@ -151,6 +155,12 @@ hooks = [
     'condition': 'checkout_win',
     'action': ['python3', 'build/util/update_midl_files.py']
   },
+  {
+    'name': 'build_libdmg_hfsplus',
+    'pattern': '.',
+    "condition": 'checkout_mac and host_os != "mac"',
+    'action': ['build/mac/cross-compile/build-libdmg-hfsplus.py', 'third_party/libdmg-hfsplus']
+  },
```
Comment (goodov) [inline DEPS:163]: I understand that this will simplify things a lot for you while working on mac installers, but I don't think we need this be enabled by default when targetting `mac` on `linux`.

Can we make all this stuff be opt-in with a gclient variable? This way you can enable it in `.gclient` via `custom_vars` section, but for other cross-compile scenarios we won't be bound with all these additional things that won't be used 99% of the time.

I'd say we should support this via `.env` variable. When enabled, it may add additional var into `.gclient` and it should enable special path manipulations you're doing in `config.js`.

Comment (mherrmann) [reply DEPS:163]: While I understand your argument, there is one subtle point that I feel it does not adequately take into account.

Your argument would make sense if there were 1,000 people or CI jobs that cross-compile. Using your number of 99%, only 10 people would then enable the `.env` or `.gclient` toggle, leaving the other 990 people/bots untouched.

But, at the moment, the total number of people and CI jobs that cross-compile is 1 - me. 100% of these people are going to set the toggle.

Yes, this number will likely be higher in the future. But even then, the number of people and jobs that do not enable the toggle will be low.

I can see a point of a `.gclient` `custom_var` that only checks out / builds the `libdmg-hfsplus` tool when enabled. That tool is only required for creating DMGs, and it incurs a dependency (CMake) that not all devs may have. If a dev just wants to cross-compile Brave, but not create a DMG, then the `custom_var` makes their life easier. Furthermore, the `custom_var` is easy to implement.

A `.env` variable that also controls the `PATH` and `PYTHONPATH` modifications however seems over the top for me, given the small numbers I mentioned above and the small effect of the path modifications.  Specifically, the path modifications do not change but only add functionality that is otherwise not available on Linux. There is no risk of Linux accidentally calling code that it expects to behave as on Linux, but it actually behaving as on macOS.

In light of this, is it OK with you if I only add the `.gclient` variable, but leave the path logic unchanged? As I said, I think it would be overkill to guard it by an entry in .env.

File: build/mac/BUILD.gn (hunk 1 above)
Comment (mherrmann) [reply build/mac/BUILD.gn]: Resolved in ac7e789b8e8a80a8623288cc34b4a6135f554345.

File: build/mac/cross-compile/path/xcrun
Comment (goodov) [inline build/mac/cross-compile/path/xcrun:1]: Are these files symlinks or shell scripts? If these are shell scripts, shouldn't they have a shebang?

File: DEPS [hunk 5]
```diff
@@ -21,6 +22,10 @@ deps = {
   "third_party/ethash/src": "https://github.com/chfast/ethash.git@e4a15c3d76dc09392c7efd3e30d84ee3b871e9ce",
   "third_party/bitcoin-core/src": "https://github.com/bitcoin/bitcoin.git@8105bce5b384c72cf08b25b7c5343622754e7337", # v25.0
   "third_party/argon2/src": "https://github.com/P-H-C/phc-winner-argon2.git@62358ba2123abd17fccf2a108a301d4b52c01a7c",
+  "third_party/libdmg-hfsplus": {
+    "url": "https://github.com/fanquake/libdmg-hfsplus.git@1cc791e4173da9cb0b0cc16c5a1aaa25d5eb5efa",
+    "condition": "checkout_dmg_tool",
```
Comment (goodov) [inline DEPS:27]: I'd keep the host/checkout_mac checks you had before (in addition to `checkout_dmg_tool`).

File: build/mac/cross-compile/path/xcrun
Comment (mherrmann) [reply build/mac/cross-compile/path/xcrun:1]: these files are symlinks to `../run-on-macos.py`.

File: build/mac/cross-compile/pythonpath/platform.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+from platform import *  # pylint: disable=wildcard-import,import-self,unused-wildcard-import
+
+sys.path = sys.path[::-1]
+sys.modules['platform'] = _this_module
+del _this_module
+del sys
+
+
+# Now change the built-in function mac_ver() so it runs on MACOS_HOST:
+def mac_ver():
```
Comment (goodov) [inline build/mac/cross-compile/pythonpath/platform.py:26]: consider storing the value for this function locally. You can do it either as a part of `npm run sync` or as a part of a build step where you replace `platform`.

you can use `run_on_macos.py` to get the required value to not duplicate `ssh` logic.

File: DEPS (hunk 4 above)
Comment (mherrmann) [reply DEPS:164]: An alternative solution I can think of, which however requires a patch, is to change upstream's `mac_ver` to read an environment variable, say `MACOS_VERSION`. Then anybody who cross-compiles can just set this to the version of macOS host version. If the env. var isn't set, then the implementation could fall back to upstream's normal implementation.

Comment (mherrmann) [reply DEPS:164]: Another alternative might be to pretend to the build process that the file `/System/Library/CoreServices/SystemVersion.plist` exists. This is where Python's `mac_ver()` reads from ([source](https://github.com/python/cpython/blob/cecaceea31f32f01b5617989e3dc8b2077f53f89/Lib/platform.py#L458C5-L458C58)).

Comment (goodov) [reply DEPS:164]: > Code signing, where upstream calls `mac_ver` in [macos_version()](https://chromium.googlesource.com/chromium/src/+/refs/heads/main/chrome/installer/mac/signing/commands.py#125). This in turn gets called by upstream in [sign_part(...)](https://chromium.googlesource.com/chromium/src/+/refs/heads/main/chrome/installer/mac/signing/signing.py#148), via [_linker_signed_arm64_needs_force(...)](https://chromium.googlesource.com/chromium/src/+/refs/heads/main/chrome/installer/mac/signing/signing.py#64).

if it's the only place, then I don't see any issues with preparing this value first via `run_on_macos.py` as part of hooks and then patch `macos_version()` (via `brave/chromium_src` or a raw patch if `brave/chromium_src` won't work).

File: DEPS (hunk 5 above)
Comment (mherrmann) [reply DEPS]: Implemented in f9e3fe97c108c7903b6dbe35fe386a944419537a.

File: DEPS (hunk 4 above)
Comment (mherrmann) [reply DEPS:164]: It might not be possible to fetch the value during hooks because the macOS host might not be reachable then.

What do you think of 3fcefc805f364998a9e6379d0ba07bc788111a4b? It requires cross-compilers to set environment variable `MACOS_VERSION`. It uses a patch but keeps the `PYTHONPATH` unchanged.

Comment (goodov) [reply DEPS:164]: is `macos_version()` required to work as a part of `//brave/build/mac/sign_app.sh`? or is it used by other GN steps that we don't own?

if we own the step(s) entirely, it would be nice to generate it automatically and only require `MACOS_HOST` for the whole thing to work, but up to you.

**Comment by goodov** [comment]
can you pls add `build/mac/cross-compile/README.md` with an explanation for all moving parts (gclient var, MACOS_HOST, `path` directory, `libdmg-hfsplus`)?

**Comment by mherrmann** [comment]
@goodov I added the README in 1e219e1f9ef6495f7b92e9e60aba6b1fdc19371e.

File: build/mac/cross-compile/README.md [hunk 7]
```diff
@@ -0,0 +1,113 @@
+# Cross-compiling Brave
+
+The code in this folder makes it possible to compile Brave for macOS in Linux.
```
Comment (goodov) [inline build/mac/cross-compile/README.md:3]: > makes it possible to compile Brave for macOS in Linux

is it really like that? I thought we can build unsigned brave and tests (mac on linux) even without this change.

Should this mention instead that it makes possible to build the signed `.app` along with valid DMG and PKG?

File: patches/chrome-installer-mac-signing-commands.py.patch [hunk 8]
```diff
@@ -0,0 +1,13 @@
+diff --git a/chrome/installer/mac/signing/commands.py b/chrome/installer/mac/signing/commands.py
+index 398a4cafbf4e46fe720f273bf8a4183d07bd4fc9..7e82a335b5c0fe8a6f3dc673d3fe14a48def5228 100644
+--- a/chrome/installer/mac/signing/commands.py
++++ b/chrome/installer/mac/signing/commands.py
+@@ -129,7 +129,7 @@ def macos_version():
+         A list containing one element for each component of the version number,
+         such as [10, 15, 6] and [11, 0].
+     """
+-    return [int(x) for x in platform.mac_ver()[0].split('.')]
++    return [int(x) for x in __import__('signing_helper').BraveGetMacRelease(platform.mac_ver()[0]).split('.')]
```
Comment (goodov) [inline patches/chrome-installer-mac-signing-commands.py.patch:10]: pls rewrite via `brave/chromium_src` file patch if possible. I think it should work here just fine.

raw patches should only be used if a less invasive option is not applicable.

File: build/mac/cross-compile/run-on-macos.py [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+exit_code = cp.returncode
+
+if cmd == 'xcodebuild' and sys.argv[1:] == ['-version']:
+    # Upstream runs this command line purely for informational purposes. If only
+    # the command line tools of Xcode are installed, then the command exits with
+    # an error (code 1). But the build goes through just fine. So ignore the
+    # non-zero exit code:
+    exit_code = 0
+
+sys.exit(exit_code)
```
Comment (goodov) [inline build/mac/cross-compile/run-on-macos.py:92]: it's hard to follow what's going on in this file. Maybe refactor this a bit to have a clear data flow?

File: build/mac/cross-compile/README.md [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+npm run build -- --target_os=mac --target_arch=x64
+```
+
+`Static` builds also work:
+
+```
+npm run build -- Static --target_os=mac --target_arch=x64
+```
+
+For `Release`, you need to disable dSYM generation and symbol stripping:
```
Comment (goodov) [inline build/mac/cross-compile/README.md:28]: > you need to disable dSYM generation and symbol stripping

why?

**Review by goodov — COMMENTED** [review]
mostly lgtm with some areas to improve

File: build/mac/cross-compile/README.md (hunk 7 above)
Comment (mherrmann) [reply build/mac/cross-compile/README.md]: True, I addressed this by modifying the README in 1d5ccac8edc32061ec03e3da0f08bbda9e6b55b5.

File: build/mac/cross-compile/run-on-macos.py (hunk 9 above)
Comment (mherrmann) [reply build/mac/cross-compile/run-on-macos.py]: Refactored in 22f99de4ce9cf53a9c90b7ba185c711aad6a0246.

File: build/mac/cross-compile/README.md (hunk 10 above)
Comment (mherrmann) [reply build/mac/cross-compile/README.md]: It doesn't "just work" and I haven't yet had time to implement this. It can be done in a follow-up PR. I added a clarifying statement about this in c825f210e9a7b930b9969004217f2064ff6e1676.

File: patches/chrome-installer-mac-signing-commands.py.patch (hunk 8 above)
Comment (mherrmann) [reply patches/chrome-installer-mac-signing-commands.py.patch]: Like in 9e0b2ce21f0ace1d2c099557983a92bfd72d66c0? `inline_chromium_src_override` was not possible because `commands.py` is not invoked directly from GN, it is copied into `out/` first. `inline_chromium_src_override` then tries to load `srv/brave/chromium_src/out/...`.

**Review by goodov — APPROVED** [review]
